jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,193 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.cloud.tools.jib.ProjectInfo;
+import com.google.cloud.tools.jib.filesystem.XdgDirectories;
+import com.google.cloud.tools.jib.json.JsonTemplate;
+import com.google.cloud.tools.jib.json.JsonTemplateMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.base.Verify;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/** Checks if Jib is up-to-date. */
+public class UpdateChecker {
+
+  /** JSON template for the configuration file used to enable/disable update checks. */
+  @VisibleForTesting
+  static class ConfigJsonTemplate implements JsonTemplate {
+    private boolean disableUpdateCheck;
+
+    @VisibleForTesting
+    void setDisableUpdateCheck(boolean disableUpdateCheck) {
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+  }
+
+  /**
+   * Begins checking for an update in a separate thread.
+   *
+   * @param versionUrl the location to check for the latest version
+   * @return a new {@link UpdateChecker}
+   */
+  public static UpdateChecker checkForUpdate(String versionUrl) {
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    Future<Optional<String>> messageFuture =
+        executorService.submit(
+            () ->
+                performUpdateCheck(
+                    Verify.verifyNotNull(ProjectInfo.VERSION), versionUrl, getConfigDir()));
+    executorService.shutdown();
+    return new UpdateChecker(messageFuture);
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    Path configFile = configDir.resolve("config.json");
+    Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+    try {
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        ConfigJsonTemplate config =
+            JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+        if (config.disableUpdateCheck) {
+          return Optional.empty();
+        }
```

Comment:
I feel like a bad configfile should be caught here -- and the user should be informed that it is indeed corrupt. We then should NOT do an update because there's no way to know if they had previously configured it or not.